compiler/src/dotty/tools/dotc/typer/Typer.scala

```diff
@@ -1977,8 +1977,6 @@ class Typer(@constructorOnly nestingLevel: Int = 0) extends Namer
               // Polymorphic SAMs are not currently supported (#6904).
               EmptyTree
           case tp =>
-            if !tp.isErroneous then
-              throw new java.lang.Error(i"internal error: closing over non-method $tp, pos = ${tree.span}")
```

Comment:
I don't think this is the correct fix. IMHO, for this kind of errors, we shouldn't run the typer in the first place.

Reply:
Hmmm, I think that this might actually be the correct fix. I even tried short-circuiting before typing ahead in Namer, but the presentation compiler leverages this for completions.

Reply:
Alright, but we should open an issue for a better error messages !